File: Proyecto 2/P2.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

struct Productos{
    int id;
    float precio;
    char nombre[50];
    struct Productos *anterior;
    struct Productos *siguiente;
};

//

struct Productos* crearNodo(int llave, float precio, char nom[]){
    struct Productos *nuevo= (struct Productos *)malloc(sizeof(struct Productos));
    if(nuevo == NULL){
        printf("Error no se puede asignar memoria\n");
        return NULL;
    }

    nuevo->id= llave;
    nuevo->precio = precio;
    strcpy(nuevo->nombre, nom);
    nuevo->anterior = NULL;
    nuevo->siguiente = NULL;

    return nuevo;
}

//

void insertarFinal(struct Productos **cabeza, int llave, int precio, char nom[]){
    struct Productos *nuevo = crearNodo(llave, precio, nom);
    struct Productos *actual;

    if(nuevo==NULL){
        return;
    }

    if(*cabeza==NULL){
        *cabeza = nuevo;
        return; 
    }

    actual = *cabeza;
    while(actual->siguiente!=NULL){
        actual = actual->siguiente;
    }

    actual->siguiente = nuevo;
    nuevo->anterior = actual;

    printf("Producto Agregado.\n");
}

//funcion buscar 

struct Productos *buscar(struct Productos *cabeza, int idBuscado){
    struct Productos *actual=cabeza;
    while(actual!=NULL){
        if(actual->id==idBuscado){
            return actual;
        }
     actual = actual->siguiente;
    }
   return NULL;

}


void deleteNode(struct Productos **head, int idDelete){
    struct Productos *eliminado = buscar(*head, idDelete);

    if (eliminado == NULL){
        printf("No se pudo eliminar el Productos ya que es inexistente");
        return;
    }
    //Caso del Productos en head
    if (eliminado->anterior == NULL){
        *head = eliminado->siguiente;
        if(*head != NULL){
            (*head)->anterior = NULL;
        }
        else {
            eliminado->anterior->siguiente = eliminado -> siguiente; 
        }
    }
    //Si el Productos final tiene un dato
    if (eliminado -> siguiente != NULL){
        eliminado -> siguiente ->anterior = eliminado -> anterior;   
    }

    free(eliminado);
    printf("Producto eliminado.\n");
}

void showListFrente(struct Productos *head){
    struct Productos *actual = head;
    if(head == NULL){
        printf("El carrito esta vacio.\n");
        return;
    }

    int i = 1;
    printf("Mostrando carrito por el frente:\n");
    while(actual != NULL){
        printf("-------Producto #%d------\n", i);
        printf("ID: %d\n", actual->id);
        printf("Nombre Producto: %s\n", actual->nombre);
        printf("Precio Producto: %f\n", actual->precio);
        actual = actual->siguiente;
        i++;
    }
    printf("\n");
}

void showListFinal(struct Productos *head){
    struct Productos *actual = head;
    if(head == NULL){
        printf("El carrito esta vacio.\n");
        return;
    }

    while(actual->siguiente != NULL){
        actual = actual->siguiente;
    }

    int i = 1;
    printf("Mostrando carrito por el final:\n");
    while(actual != NULL){
        printf("-------Producto #%d------\n", i);
        printf("ID: %d\n", actual->id);
        printf("Nombre Producto: %s\n", actual->nombre);
        printf("Precio Producto: %f\n", actual->precio);
        actual = actual->anterior;
        i++;
    }
    printf("\n");
}

float sumar(struct Productos *head){
    struct Productos *actual = head;
    if(head == NULL){
        printf("El carrito esta vacio.\n");
        return 0;
    }
    float suma = 0;
    while(actual != NULL){
        suma += actual->precio;
        actual = actual->siguiente;
    }
    return suma;
}

int main(){
    //Menu
    int opcion;
    struct Productos *llenado = NULL;
    struct Productos *encotrado;
    int id = 1;
    int idDelete;
    char nombre[50];
    float precio;
    //datos y mostrar salir
    do {
        printf("--- Carrito---\n");
        printf("1.Ingresar productos: \n");
        printf("2.Eliminar Producto: \n");
        printf("3. Mostrar Carrito Por Inicio\n");
        printf("4. Mostrar Carrito Por Final\n");
        printf("5.Salir.\n");
        printf("Seleccione una opción: ");
        scanf("%d", &opcion); 
        getchar();
        switch (opcion){
            case 1:
            printf("Ingresa Nombre del producto: ");
            fgets(nombre, 50, stdin);
            nombre[strcspn(nombre, "\n")] = '\0';
            printf("Ingresa Precio del Productos: ");
            scanf("%f", &precio);
            getchar();
            insertarFinal(&llenado, id, precio, nombre);
            id++;
            break; 

            case 2:
            showListFrente(llenado);
            printf("ID a eliminar: ");
            scanf("%d", &idDelete);
            getchar();
            deleteNode(&llenado, idDelete);
            break;
            case 3: 
            showListFrente(llenado);
            break;

            case 4:
            showListFinal(llenado);
            break;

            case 5:
            printf("Su ticket es:\n");
            showListFrente(llenado);
            float suma = sumar(llenado);
            printf("TOTAL : %.2f", suma);
            printf("Saliendo...\n");
            break;
            default:
            printf("Opción no valida\n");
        }

    } while(opcion != 5);
    
    
    return 0;
}

File: Proyecto2 Python Repaso/p2.py
import random
cantidadBase = 1

productos = {}

def ingresarProducto(name, precio):
    for j in productos:
        if productos[j]["nombre"] == name:
            productos[j]["cantidad"] += 1
            print("El producto ya estaba en el carrito, se aumenta la cantidad de este\n")
            return
    id = random.randint(100, 1000)
    productos[id] = {"nombre": name, "precioP": precio, "cantidad": cantidadBase}
    print("Producto agregado al carrito.\n")

def eliminarProducto(idDelete):
    if idDelete in productos:
        del productos[idDelete]
        print("Producto eliminado\n")
    else:
        print("ID no encontrado")

def mostrarCarritoFrente():
    if not productos:
        print("El carrito está vacío")
        return

    print(f"|\tID:\t|\tNombre\t|\tPrecio\t|\tCantidad\t|")    
    for n in productos:
        print(f"\n|\t{n}\t|\t{productos[n]["nombre"]}\t|\t{productos[n]["precioP"]}\t|\t{productos[n]["cantidad"]}\t\t|")
    
def mostrarCarritoFinal():
    if not productos:
        print("El carrito está vacío\n")
        return
    llaves_invertidas = list(productos.keys())
    llaves_invertidas.reverse() 

    print(f"|\tID:\t|\tNombre\t|\tPrecio\t|\tCantidad\t|")
    for n in llaves_invertidas:
        print(f"\n|\t{n}\t|\t{productos[n]["nombre"]}\t|\t{productos[n]["precioP"]}\t|\t{productos[n]["cantidad"]}\t\t|")

while True:
    print("\n------Carrito--------")
    print("\n1. Ingresar Producto\n2. Eliminar Producto\n3. Mostrar Carrito por el Frente\n4. Mostrar Carrito por el Final\n5. Salir\n")
    opcion = int(input("Selecciona una opción: "))
    if opcion == 1:
        nombre_Producto = input("Ingresa el Nombre del Producto: ")
        name = nombre_Producto.lower()
        precio_Producto = float(input("Ingresa el Precio del Producto: "))
        ingresarProducto(name, precio_Producto)
    elif opcion == 2:
        mostrarCarritoFrente()
        id_eliminar = int(input("Ingresa un id a eliminar: "))
        eliminarProducto(id_eliminar)
    elif opcion == 3:
        print("El carrito por el frente es:\n")
        mostrarCarritoFrente()
    elif opcion == 4:
        print("El carrito por el final es: ")
        mostrarCarritoFinal()
    elif opcion == 5:
        print("Saliendo....\n")
        break
    else:
        print("Opción inválida")
